website/docs/type-assertions.md

```diff
@@ -321,3 +329,158 @@ assertions:
   ```
 
   if it were valid in Ruby to assign to `self`.
+
+## Performance considerations
+
+Unlike `sig` annotations, type assertions *always* have a performance cost,
+even if runtime checks are globally disabled or `checked: false` is used
+at individual callsites. `T.let` and friends are ordinary Ruby method calls,
+which have intrisic overhead, in addition to the overhead of constructing
+any type arguments.
+
+This overhead isn't normally worth worrying about, but in code where you are
+already micro-optimizing to reduce method calls or object allocations, there
+are a few patterns that may be helpful:
+
+### Prefer method signatures over type assertions
+
+It's often possible to avoid using a type assertion at all, without loss of type
+safety, with a slight refactoring.
+
+For example, one common use for a type assertion is defining the type of an
+instance variable. One can frequently move this type definition to the signature
+of the constructor, taking advantage of Sorbet's ability to infer instance
+variable types when variables are set directly from constructor arguments.
+
+Instead of:
+
+```ruby
+sig {void.checked(:tests)}
+def initialize
+  @foo = T.let(MyObject.new, MyInterface)
+end
+```
+
+Write:
+
+```ruby
+sig {params(foo: MyInterface).void.checked(:tests)}
+def initialize(foo: MyObject.new)
+  @foo = foo
+end
+```
+
+In other circumstances, breaking out a method can avoid a type assertion
+(which would itself involve at least one method call anyway).
+
+For example, rather than:
+
+```ruby
+def hot_method(..)
+  # ...
+  x = T.let(polymorphic_factory(foo_please), Foo)
+  # ...
+end
+```
+
+Write:
+```ruby
+def hot_method(..)
+  # ...
+  x = make_foo
+  # ...
+end
+
+sig {returns(FooType).checked(:tests)}
+def make_foo
+  polymorphic_factory(foo_please)
+end
+```
+
+### Avoid constructing type objects
+
+The construction of an non-trivial type object is typically the most expensive
+part of a type assertion at runtime. One can usually mitigate this with the
+use of `T.type_alias`.
+
+For example, rather than:
+
+```ruby
+def hot_method(..)
+  # ...
+  foo = T.let({}, T::Hash[T.nilable(Symbol), T.any(Integer, Float)])
+  # ...
+end
+```
+
+Write:
+
+```ruby
+FooHash = T.type_alias { T::Hash[T.nilable(Symbol), T.any(Integer, Float)] }
+
+def hot_method(..)
+  # ...
+  foo = T.let({}, FooHash)
+  # ...
+end
+```
+
+### Put type assertions behind memoization
+
+Performance-sensitive methods are often memoized. In this case, it's usually
+possible to memoize the runtime type check as well.
+
+For example, rather than:
+
+```ruby
+sig {returns(Foo).checked(:tests)}
+def foo
+  @foo = T.let(@foo, T.nilable(Foo))
+  @foo ||= something_expensive
+end
+```
+
+Write:
+
+```ruby
+sig {returns(Foo).checked(:tests)}
+def foo
+  @foo ||= T.let(@foo, T.nilable(Foo))
+  @foo ||= something_expensive
+end
+```
+
+Note that for class methods, there's a better option:
```

Comment:
Given the improvements mentioned in the previous line, I'm not sure there's any meaningful difference anymore.